---
title: "Fragments and scaffolds remaining after filtering steps"
output: html_document
author: Vicky Butt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=F)
```

#### Fragments
```{r eval=FALSE}
## To run in cluster
# Number of fragments
numberOfFragments <- function(file) {
  number_of_lines <- system(paste0("wc -l ", file), intern=T)
  number_of_frag <- as.numeric(strsplit(number_of_lines, "/")[[1]][1])/4
  if(number_of_frag%%1 != 0){
    stop("Number of lines not divisible by 4. Is the fastq file complete?")
  }
  return(number_of_frag)
}

## Set samples and work directory
args = commandArgs(trailingOnly=TRUE)
sample <- as.character(read.table(args[1])[1,1])
folder_work <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/DATA_RAW"

## Number of fragments before trim
file <- paste0(folder_work, "/", sample, "_1.fastq")
frag_before_trim <- numberOfFragments(file)

## Number of fragments after trim
file <- paste0(folder_work, "/", sample, "_trimmo_trimmed_1.fastq")
frag_after_trim <- numberOfFragments(file)

## Number of fragments after bowtie filter
file <- paste0(folder_work, "/", sample, "_trimmo_trimmed_and_filtered_1.fastq")
frag_after_filter <- numberOfFragments(file)

# Save stats
stats <- c(frag_before_trim, frag_after_trim, frag_after_filter)/frag_before_trim*100
save(stats, file = paste0(folder_work, "/fragment_stats_", sample, ".RData"))
```

```{r fragment_heatmap}
# Visualisation on PC
library("RColorBrewer")
library(gplots)
setwd("/Users/victoriabutt/Documents/MS/scripts")

# List stats data files in directory
data_directory <- "RData/fragment_stats/"
file_list <- list.files(paste0(getwd(), "/", data_directory))

# Extract sample names
samples <- gsub("fragment_stats_", "", gsub(".RData", "", file_list))

# Create data frame
df_stats <- c()
for(i in 1:length(file_list)){
  load(paste0(data_directory,file_list[i]))
  df_stats <- rbind(df_stats, stats)
}
row.names(df_stats) <- samples
colnames(df_stats) <- c("Before trim", "After trim", "After \n bowtie filter")

# Create heatmap
heatmap.2(df_stats, Rowv = FALSE, Colv = FALSE, dendrogram = 'none',
          scale = 'none', trace = 'none',
          col = brewer.pal(9, "Blues"), cexCol=1, cexRow=0.8, srtCol = 45,  
          key.title = "", key.xlab = "% fragments remaining",
          rowsep = seq(1,100,1), colsep = seq(1,100,1),
          sepwidth = c(0.005, 0.005), main = "Fragments remaining\nafter filtering steps")
```

#### Scaffolds
```{r eval=FALSE}
## To run in cluster
# Number of scaffolds
numberOfScaffolds <- function(file) {
  number_of_scaffolds <- as.numeric(system(paste0('grep -c "^>" ', file), intern=T))
  return(number_of_scaffolds)
}

## Set samples and work directory
args = commandArgs(trailingOnly=TRUE)
sample <- as.character(read.table(args[1])[1,1])
folder_filter_scaffold <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/FILTERED_SCAFFOLDS"

## Number of fragments before bowtie filter
file <- paste0(folder_filter_scaffold, "/", sample, "_scaffolds.fasta")
scaf_before_filter <- numberOfScaffolds(file)

## Number of fragments after bowtie filter
file <- paste0(folder_filter_scaffold, "/", sample, "_scaffolds_filtered.fasta")
scaf_after_filter <- numberOfScaffolds(file)

# Save stats
stats <- c(scaf_before_filter, scaf_after_filter)/scaf_before_filter*100
save(stats, file = paste0(folder_filter_scaffold, "/scaffold_stats_", sample, ".RData"))
```

```{r scaffolds_heatmap}
# Visualisation on PC
library("RColorBrewer")
library(gplots)
setwd("/Users/victoriabutt/Documents/MS/scripts")

# List stats data files in directory
data_directory <- "RData/scaffold_stats/"
file_list <- list.files(paste0(getwd(), "/", data_directory))

# Extract sample names
samples <- gsub("scaffold_stats_", "", gsub(".RData", "", file_list))

# Create data frame
df_stats <- c()
for(i in 1:length(file_list)){
  load(paste0(data_directory,file_list[i]))
  df_stats <- rbind(df_stats, stats)
}
row.names(df_stats) <- samples
colnames(df_stats) <- c("Before filter", "After filter")

# Create heatmap
heatmap.2(df_stats, Rowv = FALSE, Colv = FALSE, dendrogram = 'none',
          scale = 'none', trace = 'none',
          col = brewer.pal(9, "Blues"), cexCol=1, cexRow=0.8, srtCol = 45,  
          key.title = "", key.xlab = "% scaffolds remaining",
          rowsep = seq(1,100,1), colsep = seq(1,100,1),
          sepwidth = c(0.005, 0.005), main = "Scaffolds remaining\nafter filtering step")
```
