---
title: "Supplementary Code"
output: html_document
author: Vicky Butt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

```{r}
# Download files
setwd("/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/DATA_RAW")

file_to_download <- 
  c("ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX023/SRX023430/SRR060075/SRR060075.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX023/SRX023430/SRR060076/SRR060076.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX023/SRX023406/SRR060040/SRR060040.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX023/SRX023406/SRR064436/SRR064436.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX023/SRX023489/SRR060154/SRR060154.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX024/SRX024082/SRR061324/SRR061324.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX024/SRX024082/SRR061359/SRR061359.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX024/SRX024164/SRR061490/SRR061490.sra",
    "ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX024/SRX024164/SRR061492/SRR061492.sra")

file_to_process <- paste0("/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/DATA_RAW/", c("SRR060075", "SRR060076", "SRR060040", "SRR064436", "SRR060154", "SRR061324", "SRR061359", "SRR061490", "SRR061492"))

file_to_process_unique_FINAL_Rosalind <- cbind(file_to_download, file_to_process)

for(i in 1:nrow(file_to_process_unique_FINAL_Rosalind)){
 # download
 order <- paste("wget ",as.character(file_to_process_unique_FINAL_Rosalind[i,1]),sep="")
 system(order)
 
 # aspera
 order2 <- paste("/users/k1639482/brc_scratch/SOFTWARE/sratoolkit.2.8.1-3-centos_linux64/bin/fastq-dump  -I --split-files ",
                 as.character(file_to_process_unique_FINAL_Rosalind[i,2]),".sra",sep="")
 system(order2)
 
 # delete sra
 order3 <- paste("rm ", as.character(file_to_process_FINAL_test_Rosalind[i,2]),".sra",sep="")
 system(order3)
}
```

### Run multi-threaded batch job
```{r}
# #!/bin/bash
# 
# #$ -o /users/k1639482/brc_scratch/grid_schedular_outputs/sge-out
# #$ -e /users/k1639482/brc_scratch/grid_schedular_outputs/sge-out
# 
# #$ -q HighMemLongterm.q,HighMemShortterm.q
# 
# #$ -hold_jid_ad 093966 -t 1-8 
# 
# echo input.$SGE_TASK_ID.txt
# module load general/R/3.2.1
# Rscript <R file> /users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/input_files/input.$SGE_TASK_ID.txt
```

```{r}
# Set working directories and samples
args = commandArgs(trailingOnly=TRUE)
sample <- as.character(read.table(args[1])[1,1])
folder_work <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/DATA_RAW"
folder_spades <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/SPADES"
folder_filter_scaffold <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/FILTERED_SCAFFOLDS"
```

### Step 1: Trim raw reads with AlienTrimmer 0.4.0
Documentation: http://gensoft.pasteur.fr/docs/AlienTrimmer/0.4.0/AlienTrimmer.v.0.4.doc.pdf

```{r alientrimmer}
# ADAPTERS INFORMATION
# http://bioinformatics.cvr.ac.uk/blog/illumina-adapter-and-primer-sequences/
# http://support.illumina.com/downloads/illumina-customer-sequence-letter.html
# http://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences_1000000002694-01.pdf

alientrimmer <- "/users/k1639482/brc_scratch/SOFTWARE/AlienTrimmer_0.4.0/src/AlienTrimmer.jar"
parameters_AT <- " -k 10 -l 45 -m 5 -p 40 -q 20 "
# Only p changed from default 0

set_primers <- "/users/k1639482/brc_scratch/NECESSARY_FILES/alienTrimmerPF8contaminants.fasta"

order <- c("java -jar -d64 -Xmx512M")
order <- paste(order, " " ,alientrimmer," ", parameters_AT,
               " -if ",folder_work,"/",sample,"_1.fastq ",
               " -ir ",folder_work,"/",sample,"_2.fastq ",
               " -c ",set_primers," ",
               " -of ",folder_work,"/",sample,"_trimmo_trimmed_1.fastq",
               " -or ",folder_work,"/",sample,"_trimmo_trimmed_2.fastq",
               " -os ",folder_work,"/",sample,"_trimmo_trimmed_s.fastq",
               sep="")
system(order)
```

### Step 2: Remove human DNA using Bowtie2 2.2.3
Documentation: http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml

```{r bowtie}
# screen -r HMP_HUMAN_FILTER boogy5
# Version: Bowtie2-2.2.3

bowtie2 <- "/users/k1639482/brc_scratch/SOFTWARE/bowtie2-2.2.3/bowtie2"
max_processors <- 40
reference_human <- "/users/k1639482/brc_scratch/NECESSARY_FILES/bowtie2/homo_sapiens.fna"
samples_REF <- c("homo_sapiens.fna.1.bt2", "homo_sapiens.fna.2.bt2", "homo_sapiens.fna.3.bt2", 
                 "homo_sapiens.fna.4.bt2", "homo_sapiens.fna.rev.1.bt2", "homo_sapiens.fna.rev.2.bt2")
parameters_BT <- paste(" -q -N 1 -p  ",max_processors,"  -k 1 --fr -x ",reference_human,
                       " --end-to-end --phred33 --very-sensitive --no-discordant ",sep="")

order <- c(bowtie2)
order <- paste(order, " ", parameters_BT,
               " -1 ",folder_work,"/",sample,"_trimmo_trimmed_1.fastq ",
               " -2 ",folder_work,"/",sample,"_trimmo_trimmed_2.fastq ",
               " -S /dev/null", # ",folder_work,"/",sample,"_trimmo_trimmed_human",".sam ",
               " --un-conc ",folder_work,"/",sample,"_trimmo_trimmed_and_filtered","_%.fastq",sep="")
print("Running bowtie2....")
system(order)
```


### Step 3: Additional filtering by assembly and alignment
Documentation: http://spades.bioinf.spbau.ru/release3.9.0/manual.html
Documentation: http://www.htslib.org/doc/samtools.html

```{r filter_scaffolds}
## Assemble reads with SPAdes 3.9.0
spades <- "/users/k1639482/brc_scratch/SOFTWARE/SPAdes-3.9.0-Linux/bin/spades.py"

parameters <- " -t 24 -m 220  -k 21,33,55 --only-assembler --meta "

order <- c(spades)
order <- paste(order, parameters,
               " -1 ",folder_work, "/", sample,"_trimmo_trimmed_and_filtered_1.fastq",
               " -2 ",folder_work, "/", sample,"_trimmo_trimmed_and_filtered_2.fastq", 
               " -o ",folder_spades,"/",sample," ",
               sep="")
system(order)

## Additional filter step with Bowtie2 2.2.3
bowtie2 <- "/users/k1639482/brc_scratch/SOFTWARE/bowtie2-2.2.3"
max_processors <- 32

# Copy scaffold to filtered scaffold folder
order_cp_scaffold <- paste("cp ",folder_spades,"/",sample,"/scaffolds.fasta ",
                           folder_filter_scaffold,"/",sample,"_scaffolds.fasta",
                           sep="")
system(order_cp_scaffold)

# Generate bowtie2 reference
order_bowtieindex_scaffold <- paste(bowtie2, "/", "bowtie2-build -f ",
                                    folder_filter_scaffold,"/",sample,"_scaffolds.fasta ",
                                    folder_filter_scaffold, "/", sample,"_scaffolds",
                                    sep="")
system(order_bowtieindex_scaffold)

# Align trimmed/filtered reads to bowtie2 reference
parameters_BT_go <- paste(" -q -N 1 -p  ",max_processors," -k 1 --fr",
                          " -x ", folder_filter_scaffold,"/",sample,"_scaffolds ",
                          " --end-to-end --phred33 --very-sensitive --no-discordant ",sep="")
order_bt_scaffold <- paste(bowtie2, "/", "bowtie2 ", parameters_BT_go,
                           " -1 ",folder_work,"/",sample,"_trimmo_trimmed_and_filtered_1.fastq ",
                           " -2 ",folder_work,"/",sample,"_trimmo_trimmed_and_filtered_2.fastq ",
                           " -S ",folder_filter_scaffold,"/",sample,"_scaffold.sam",sep="")
system(order_bt_scaffold)

## SAMtools
samtools <- "/opt/apps/bioinformatics/samtools/1.3.1/bin/"

order_sam_to_bam <- paste(samtools, "samtools view -bS ", 
                          folder_filter_scaffold,"/",sample,"_scaffold.sam ",
                          " > ", folder_filter_scaffold,"/",sample,"_scaffold.bam ",
                          sep="")
system(order_sam_to_bam)
order_sort <- paste(samtools, "samtools sort ", folder_filter_scaffold,"/",sample,"_scaffold.bam ",
                    " -o ", folder_filter_scaffold,"/",sample,"_scaffold_sort.bam",
                    sep="")
system(order_sort)
order_index <- paste(samtools, "samtools index -b ", 
                     folder_filter_scaffold,"/",sample,"_scaffold_sort.bam ",
                     sep="")
system(order_index)

# ## delete sam files and unsorted bam files
# rm_1<- paste("rm ",folder_filter_scaffold,"/",sample,"_scaffold.sam",sep="" )
# rm_2<- paste("rm ",folder_filter_scaffold,"/",sample,"_scaffold.bam",sep="" )
# system(rm_1)
# system(rm_2)

## Filter
#samtools idxstats in.sam|in.bam|in.cram
#Retrieve and print stats in the index file corresponding to the input file. Before calling idxstats, the input BAM file must be indexed by samtools index.
#The output is TAB-delimited with each line consisting of reference sequence name, sequence length, # mapped reads and # unmapped reads. It is written to stdout.
order_filter <- paste(samtools, "samtools idxstats ", 
                      folder_filter_scaffold,"/",sample,"_scaffold_sort.bam > ",
                      folder_filter_scaffold,"/",sample,"_scaffold_sort_quant.txt ",
                      sep="")

system(order_filter)

require(Biostrings)
# Read the summary file
filter_info <- read.csv(paste(folder_filter_scaffold,"/",sample,"_scaffold_sort_quant.txt",sep=""),
                        sep="\t",row.names=1,header=F)
filter_info <- filter_info[rownames(filter_info) != "*",]

# Remove scaffolds shorter than 500 and gene length less than 11
S1 <- filter_info[,1] >=500 & filter_info[,2] >10
names_kept <- rownames(filter_info)[S1]

# Read the FASTA file and filter
object_use <- readDNAStringSet(paste(folder_filter_scaffold,"/",sample,"_scaffolds.fasta",sep="") )
object_use_filter <- object_use[names(object_use) %in% names_kept]

# Write FASTA file
writeXStringSet(object_use_filter, paste(folder_filter_scaffold,"/",sample,"_scaffolds_filtered.fasta",sep=""), append=FALSE,
                compress=FALSE, compression_level=NA, format="fasta")
```

### Step 4: Quality Control

##### Fragments
```{r eval=FALSE}
## To run in cluster
# Number of fragments
numberOfFragments <- function(file) {
  number_of_lines <- system(paste0("wc -l ", file), intern=T)
  number_of_frag <- as.numeric(strsplit(number_of_lines, "/")[[1]][1])/4
  if(number_of_frag%%1 != 0){
    stop("Number of lines not divisible by 4. Is the fastq file complete?")
  }
  return(number_of_frag)
}

## Set samples and work directory
args = commandArgs(trailingOnly=TRUE)
sample <- as.character(read.table(args[1])[1,1])
folder_work <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/DATA_RAW"

## Number of fragments before trim
file <- paste0(folder_work, "/", sample, "_1.fastq")
frag_before_trim <- numberOfFragments(file)

## Number of fragments after trim
file <- paste0(folder_work, "/", sample, "_trimmo_trimmed_1.fastq")
frag_after_trim <- numberOfFragments(file)

## Number of fragments after bowtie filter
file <- paste0(folder_work, "/", sample, "_trimmo_trimmed_and_filtered_1.fastq")
frag_after_filter <- numberOfFragments(file)

# Save stats
stats <- c(frag_before_trim, frag_after_trim, frag_after_filter)
save(stats, file = paste0(folder_work, "/fragment_stats_", sample, ".RData"))
```

##### Scaffolds
```{r eval=FALSE}
## To run in cluster
# Number of scaffolds
numberOfScaffolds <- function(file) {
  number_of_scaffolds <- as.numeric(system(paste0('grep -c "^>" ', file), intern=T))
  return(number_of_scaffolds)
}

## Set samples and work directory
args = commandArgs(trailingOnly=TRUE)
sample <- as.character(read.table(args[1])[1,1])
folder_filter_scaffold <- "/users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/FILTERED_SCAFFOLDS"

## Number of fragments before bowtie filter
file <- paste0(folder_filter_scaffold, "/", sample, "_scaffolds.fasta")
scaf_before_filter <- numberOfScaffolds(file)

## Number of fragments after bowtie filter
file <- paste0(folder_filter_scaffold, "/", sample, "_scaffolds_filtered.fasta")
scaf_after_filter <- numberOfScaffolds(file)

# Save stats
stats <- c(scaf_before_filter, scaf_after_filter)
save(stats, file = paste0(folder_filter_scaffold, "/scaffold_stats_", sample, ".RData"))
```

### Step 5: Gene Predictions
Prodigal documentation: https://github.com/hyattpd/prodigal/wiki
\
CD-HIT documentation: http://weizhongli-lab.org/lab-wiki/doku.php?id=cd-hit-user-guide
```{r}
## Bash for merging files

# #!/bin/bash
# 
# #$ -o users/k1639482/brc_scratch/grid_schedular_outputs/sge-out
# #$ -e users/k1639482/brc_scratch/grid_schedular_outputs/sge-out
# 
# 
# #$ -q HighMemLongterm.q,HighMemShortterm.q
# 
# cat /users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/FILTERED_SCAFFOLDS/*_scaffolds_filtered.fasta > /users/k1639482/brc_scratch/ORAL_MICROBIOME/HMP/FILTERED_SCAFFOLDS/merged_scaffolds_filtered.fasta

sample <- "merged"
folder_py <- "/users/k1639482/brc_scratch/SOFTWARE/source_py"
prodigal <- "/opt/apps/bioinformatics/prodigal/2.6.3/prodigal"
require(Biostrings)

## Prodigal
# http://prodigal.ornl.gov
# https://github.com/hyattpd/prodigal/wiki

# Anonymous Mode, in which Prodigal applies pre-calculated training files to the 
# provided input sequence and predicts genes based on the best results. (for metagenomics)

#Usage:  prodigal [-a trans_file] [-c] [-d nuc_file] [-f output_type]
#                 [-g tr_table] [-h] [-i input_file] [-m] [-n] [-o output_file]
#                 [-p mode] [-q] [-s start_file] [-t training_file] [-v]

#         -a:  Write protein translations to the selected file.
#         -c:  Closed ends.  Do not allow genes to run off edges.
#         -d:  Write nucleotide sequences of genes to the selected file.
#         -f:  Select output format (gbk, gff, or sco).  Default is gbk.
#         -g:  Specify a translation table to use (default 11).
#         -h:  Print help menu and exit.
#         -i:  Specify FASTA/Genbank input file (default reads from stdin).
#         -m:  Treat runs of N as masked sequence; don't build genes across them.
#         -n:  Bypass Shine-Dalgarno trainer and force a full motif scan.
#         -o:  Specify output file (default writes to stdout).
#         -p:  Select procedure (single or meta).  Default is single.
#         -q:  Run quietly (suppress normal stderr output).
#         -s:  Write all potential genes (with scores) to the selected file.
#         -t:  Write a training file (if none exists); otherwise, read and use
#              the specified training file.
#         -v:  Print version number and exit.

order_prodigal <- paste(prodigal, " -i ",folder_filter_scaffold,"/",sample,"_scaffolds_filtered.fasta ",
                        "-a ", folder_filter_scaffold,"/",sample,"_prot.faa ",
                        "-d ", folder_filter_scaffold,"/",sample,"_gene.fna ",
                        " -p meta ",
                        sep="")
system(order_prodigal)

# Filter the contigs and fragments less than 60
order_filter_gene_length <- paste("python ",folder_py,"/FilterFasta/FilterFasta.py ",
                                  "-f ",folder_filter_scaffold,"/",sample,"_gene.fna ",
                                  " -t 60 ",
                                  " -o ",folder_filter_scaffold,"/",sample,"_gene_filt",60,".fna ",sep="")
system(order_filter_gene_length)

# Remove incomplete set of fragment in a prodigal file, i.e. remove those of prodigal choice 11 (out of 4 and 11), genetic code 4 only associated with fungi
order_filter_both_ends <- paste("python ",folder_py,"/FilterProdigal/FilterProdigal.py ",
                                " -i ",folder_filter_scaffold,"/",sample,"_gene_filt",60,".fna ",
                                " -c 11 ",
                                " -o ",folder_filter_scaffold,"/",sample,"_input_gene.fna ",sep="")
system(order_filter_both_ends)

# Read the FASTA file
object_use <- readDNAStringSet(paste(folder_filter_scaffold,"/",sample,"_input_gene.fna",sep="") )
object_use_names <- names(object_use)
for(h in 1:length(object_use_names))
{
  object_use_names[h] <- strsplit(object_use_names[h],split=" ")[[1]][1]
}
write.table(object_use_names,paste(folder_filter_scaffold,"/",sample,"_interest_list.txt",sep=""),
            quote=F,row.names=F,col.names=F,sep="")

# Filter proteins based on selected genes
order_filter_protein <- paste("python ",folder_py,"/ExtractProteins/ExtractProteins.py ",
                              "-i ",folder_filter_scaffold,"/",sample,"_interest_list.txt ",
                              " -d ",folder_filter_scaffold,"/",sample,"_prot.faa ",
                              " -o ",folder_filter_scaffold,"/",sample,"_prot_filt.faa ",sep="")
system(order_filter_protein)

## Cluster genes with cd-hit
cd_hit <- "/opt/apps/bioinformatics/cd-hit/4.6.6/cd-hit-est"
order_cd_hit <- paste(cd_hit,
                      " -i ",folder_filter_scaffold,"/",sample,"_input_gene.fna ",
                      " -o ",folder_filter_scaffold,"/",sample,"_gene_filt",60,"_cdhit.fna ",
                      "  -aS 0.9 -c 0.95 -T 0 -M 0 -t 0 -d 0 -G 0 ",sep="")
system(order_cd_hit)
```